docs/LibraryEvolution.md

```diff
@@ -0,0 +1,763 @@
+> in versions. Our generalized switch patterns probably make it hard to
+> nail down the behavior here.
+
+Inlineable Functions
+--------------------
+
+Functions are a very common example of resilience: the function's
+declaration is published as API, but its body may change between library
+versions as long as it upholds the same semantic contracts. This applies
+to other function-like constructs as well: initializers, accessors, and
+deinitializers.
+
+However, sometimes it is useful to provide the body to clients as well.
+There are a few common reasons for this:
+
+-   The function only performs simple operations, and so inlining it
+    will both save the overhead of a cross-library function call and
+    allow further optimization of callers.
+-   The function accesses a fixed-layout struct with non-public members;
+    this allows the library author to preserve invariants while still
+    allowing efficient access to the struct.
+
+A public function marked with the `@inlineable` attribute makes its body
+available to clients as part of the module's public interface. The
+`@inlineable` attribute takes a version number, just like `@available`;
+clients may not assume that the body of the function is suitable when
+deploying against older versions of the library.
+
+Clients are not required to inline a function marked `@inlineable`.
+
+> Note: It is legal to change the implementation of an inlineable function in
+> the next release of the library. However, any such change must be made
+> with the understanding that it may or may not affect existing clients.
+
+### Restrictions
+
+Because the body of an inlineable function (or method, accessor,
+initializer, or deinitializer) may be inlined into another module, it
+must not make any assumptions that rely on knowledge of the current
+module. Here is a trivial example:
+
+```swift
+public struct Point2D {
+  var x, y: Double
+  public init(x: Double, y: Double) { … }
+}
+
+extension Point2D {
+  @inlineable public func distanceTo(other: Point2D) -> Double {
+    let deltaX = self.x - other.x
+    let deltaY = self.y - other.y
+    return sqrt(deltaX*deltaX + deltaY*deltaY)
+  }
+}
+```
+
+As written, this `distanceTo` method is not safe to inline. The next
+release of the library could very well replace the implementation of
+`Point2D` with a polar representation:
+
+```swift
+public struct Point2D {
+  var r, theta: Double
+  public init(x: Double, y: Double) { … }
+}
+```
+
+and the `x` and `y` properties have now disappeared. To avoid this, we
+have the following restrictions on the bodies of inlineable functions:
+
+-   **They may not define any local types** (other than typealiases).
+-   **They must not reference any** `private` **entities,** except for
+    local functions declared within the inlineable function itself.
+-   **They must not reference any** `internal` **entities except for
+    those that have been** [availability-pinned](#pinning). See below
```

Comment:
This isn't your fault, but the bold here should really go across the link as well. ReST just doesn't allow nested markup.